boa/src/syntax/ast/node.rs

```diff
@@ -0,0 +1,300 @@
+    /// Gets the constant field of a value
+    GetConstField(Box<Node>, String),
+    /// Gets the [field] of a value
+    GetField(Box<Node>, Box<Node>),
+    /// [init], [cond], [step], body
+    ForLoop(
+        Option<Box<Node>>,
+        Option<Box<Node>>,
+        Option<Box<Node>>,
+        Box<Node>,
+    ),
+    /// Check if a conditional expression is true and run an expression if it is and another expression if it isn't
+    If(Box<Node>, Box<Node>, Option<Box<Node>>),
+    /// Let declaraton
+    LetDecl(Vec<(String, Option<Node>)>),
+    /// Load a reference to a value, or a function argument
+    Local(String),
+    /// New
+    New(Box<Node>),
+    /// Create an object out of the binary tree given
+    ObjectDecl(Box<BTreeMap<String, Node>>),
+    /// Object Declaration (replaces ObjectDecl)
+    Object(Vec<PropertyDefinition>),
+    /// Return the expression from a function
+    Return(Option<Box<Node>>),
+    /// Run blocks whose cases match the expression
+    Switch(Box<Node>, Vec<(Node, Vec<Node>)>, Option<Box<Node>>),
+    /// Spread operator
+    Spread(Box<Node>),
+    // Similar to Block but without the braces
+    StatementList(Vec<Node>),
+    /// Throw a value
+    Throw(Box<Node>),
+    /// Return a string representing the type of the given expression
+    TypeOf(Box<Node>),
+    /// Try / Catch
+    Try(
+        Box<Node>,
+        Option<Box<Node>>,
+        Option<Box<Node>>,
+        Option<Box<Node>>,
+    ),
+    This,
+    /// Run an operation on a value
+    UnaryOp(UnaryOp, Box<Node>),
+    /// A variable declaration
+    VarDecl(Vec<(String, Option<Node>)>),
+    /// Repeatedly run an expression while the conditional expression resolves to true
+    WhileLoop(Box<Node>, Box<Node>),
+}
+
+impl Operator for Node {
+    fn get_assoc(&self) -> bool {
+        match *self {
+            Node::Construct(_, _)
+            | Node::UnaryOp(_, _)
+            | Node::TypeOf(_)
+            | Node::If(_, _, _)
+            | Node::Assign(_, _) => false,
+            _ => true,
+        }
+    }
+    fn get_precedence(&self) -> u64 {
+        match self {
+            Node::GetField(_, _) | Node::GetConstField(_, _) => 1,
+            Node::Call(_, _) | Node::Construct(_, _) => 2,
+            Node::UnaryOp(UnaryOp::IncrementPost, _)
+            | Node::UnaryOp(UnaryOp::IncrementPre, _)
+            | Node::UnaryOp(UnaryOp::DecrementPost, _)
+            | Node::UnaryOp(UnaryOp::DecrementPre, _) => 3,
+            Node::UnaryOp(UnaryOp::Not, _)
+            | Node::UnaryOp(UnaryOp::Tilde, _)
+            | Node::UnaryOp(UnaryOp::Minus, _)
+            | Node::TypeOf(_) => 4,
+            Node::BinOp(op, _, _) => op.get_precedence(),
+            Node::If(_, _, _) => 15,
+            // 16 should be yield
+            Node::Assign(_, _) => 17,
+            _ => 19,
+        }
+    }
+}
+
+impl Display for Node {
+    fn fmt(&self, f: &mut Formatter) -> Result {
```

Comment:
We should probably add some indentation and proper formatting here, as I did in #263. Maybe I can port those changes to this branch, if it's ok?

Reply:
Sure!